visionarypm/__init__.py

```diff
@@ -106,35 +103,86 @@ def getConfig():
 
 
 def generate(master_password, keyword, cost, oLen=None, num_words=None):
+
     if not num_words:
         num_words = params['nwords']
-    dict_len = len(words)
-    entropy_per_word = math.log(dict_len, 2)
-    hash = codecs.encode(scrypt.hash(password=master_password, salt=keyword, N=1<<cost, buflen=64), 'hex').decode('utf-8')
-    available_entropy = len(hash) * 4
-    seed = int(hash, 16)
-    if (num_words * entropy_per_word) > available_entropy:
-        raise Exception("The output entropy of the specified hashfunc (%d) is too small." % available_entropy)
-    phrase = []
-    for i in range(num_words):
-        remainder = seed % dict_len
-        seed = seed // dict_len
-        phrase.append(words[int(remainder)].strip().decode('utf-8'))
-    readable_hash = " ".join(phrase).lower().capitalize()
+
+    def normal_password():
+        seed = scrypt.hash(password=master_password,
+                           salt=keyword,
+                           N=1<<cost,
+                           buflen=64)
+        return codecs.encode(seed, 'hex').decode('utf-8')
+
+    normal = normal_password()
+
+    # Courtesy of rubik
+    def complex_password():
+        # A random shuffle of string.punctuation + string.ascii_uppercase
+        to_add = '*>OI+D~}E{AF"`KXR\[J:-&#BQWN;$%TGPZ.=U,_HS()@L\'|<M]YCV/?!^'
```

Comment:
`to_add` contains all punctuation characters and all uppercase letters. It was generated as follows:

    to_add = list(string.punctuation + string.ascii_uppercase)
    random.shuffle(to_add)
    to_add = ''.join(to_add)

In theory, there is no difference between the original `to_add` and the shuffled `to_add`, but the latter just gives better-looking passwords. In the other case we would get either too many punctuation characters or too many uppercase letters.

An improvement could be to compute the `to_add` string at the first start of the program and include it in the configuration file with the cost factor and password length.